### DISCUSSION TITLE
MySQL 索引性能分析概要 · /sql-index-performance

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#248] [Announcements]
https://draveness.me/sql-index-performance

**Comment by levicc** [comment]
1. 问下博主辅助索引重复值是怎么存的，我看了B+树，对于重复值是用了溢出页。做个假设一页存100条数据，那么600条需要存6页，而且这6页并不一定是物理连续的，可能会在不同的磁道上，那么是不是要在原有的1次基础上加上6次？
2. 辅助索引里为什么要存主键，不直接存对应数据的物理地址，这样的话就少了主键索引B+树的高度。
3. 索引的查找应该是B+树高度次随机访问吧？

**Comment by draveness** [comment]
> @levicc
> 1. 问下博主辅助索引重复值是怎么存的，我看了B+树，对于重复值是用了溢出页。做个假设一页存100条数据，那么600条需要存6页，而且这6页并不一定是物理连续的，可能会在不同的磁道上，那么是不是要在原有的1次基础上加上6次？
> 2. 辅助索引里为什么要存主键，不直接存对应数据的物理地址，这样的话就少了主键索引B+树的高度。
> 3. 索引的查找应该是B+树高度次随机访问吧？

1. 第一个问题不是很明白你问的是什么
2. 索引里存的不一定是主键，一般应该叫做 bookmark（忘记细节了），在 MySQL 里是主键，第三个回答应该能回答后半部分的问题吧
3. 索引占用的磁盘空间非常小，一亿数据的表（索引行平均长度 100 字节）全部的索引加起来大概也只有 400MB，都加在到内存中也不是什么问题，访问索引中的数据大多数需要 10-20ms 的时间，一般就是 1-2 次随机访问的时间，而不是树的高度
    + 所以主键索引大部分也会在缓存中，在**大多数时候**主键就可以理解为数据的物理地址，就算直接存数据对应的物理地址，也仍然需要一次随机访问获取整条记录。

**Comment by levicc** [comment]
> @Draveness
> > @levicc
> > 1. 问下博主辅助索引重复值是怎么存的，我看了B+树，对于重复值是用了溢出页。做个假设一页存100条数据，那么600条需要存6页，而且这6页并不一定是物理连续的，可能会在不同的磁道上，那么是不是要在原有的1次基础上加上6次？
> > 2. 辅助索引里为什么要存主键，不直接存对应数据的物理地址，这样的话就少了主键索引B+树的高度。
> > 3. 索引的查找应该是B+树高度次随机访问吧？
> 
> 1. 第一个问题不是很明白你问的是什么
> 2. 索引里存的不一定是主键，一般应该叫做 bookmark（忘记细节了），在 MySQL 里是主键，第三个回答应该能回答后半部分的问题吧
> 3. 索引占用的磁盘空间非常小，一亿数据的表（索引行平均长度 100 字节）全部的索引加起来大概也只有 400MB，都加在到内存中也不是什么问题，访问索引中的数据大多数需要 10-20ms 的时间，一般就是 1-2 次随机访问的时间，而不是树的高度
> 

后面两个问题明白了，第一个问题是，你上面那个例子，扫描索引有600条数据。因为一页大概存2-200条数据，那么600条数据肯定是在不同的页里，这些页不一定是物理相邻的，为什么可以顺序读取呢？感觉我纠结的点在于，顺序读取是指的这些页在物理空间上连续吗？

**Comment by draveness** [comment]
> @levicc
> > @Draveness
> > > @levicc
> > > 1. 问下博主辅助索引重复值是怎么存的，我看了B+树，对于重复值是用了溢出页。做个假设一页存100条数据，那么600条需要存6页，而且这6页并不一定是物理连续的，可能会在不同的磁道上，那么是不是要在原有的1次基础上加上6次？
> > > 2. 辅助索引里为什么要存主键，不直接存对应数据的物理地址，这样的话就少了主键索引B+树的高度。
> > > 3. 索引的查找应该是B+树高度次随机访问吧？
> > 
> > 1. 第一个问题不是很明白你问的是什么
> > 2. 索引里存的不一定是主键，一般应该叫做 bookmark（忘记细节了），在 MySQL 里是主键，第三个回答应该能回答后半部分的问题吧
> > 3. 索引占用的磁盘空间非常小，一亿数据的表（索引行平均长度 100 字节）全部的索引加起来大概也只有 400MB，都加在到内存中也不是什么问题，访问索引中的数据大多数需要 10-20ms 的时间，一般就是 1-2 次随机访问的时间，而不是树的高度
> > 
> 
> 后面两个问题明白了，第一个问题是，你上面那个例子，扫描索引有600条数据。因为一页大概存2-200条数据，那么600条数据肯定是在不同的页里，这些页不一定是物理相邻的，为什么可以顺序读取呢？感觉我纠结的点在于，顺序读取是指的这些页在物理空间上连续吗？

这些页在物理空间上并不一定连续，但是在计算索引性能时为了简化计算的过程，一般不考虑这部分耗时。
